### ISSUE TITLE
maildir: mishandling of messages moved/deleted by external tools (e.g., imapsync)

### ISSUE THREAD

**Issue by rapha8l** [neomutt/neomutt#2077]
Hi,

Since a few releases (sorry I can say exactly when it stopped to work), mails moved by imapfilter in other folders are still visible in inbox index, though not here anymore, and also can't be opened.

They move from the index only when navigating in other folders

Thanks

**Comment by jindraj** [comment]
similar issue [Message externally deleted from Maildir no longer disappears from index](#2150)

**Comment by gahr** [comment]
for the record, the IMAP backend handles emails deleted by another client just "fine"
* I get the usual  "Mailbox was externally modified. Flags may be wrong." messsage
* The selection moves to a random message in the index (usually one from several years ago)
* All threads are expanded

**Comment by gahr** [comment]
Symptoms:

Changes (moves / deletes) done via another client, then sync'd to local maildir by offlineimap/imapfilter, are not picked up by NeoMutt.
From NeoMutt's index, the messages are still visible in their own folders, but not accessible. Trying do do anything with them results in `No such file or directory (errno = 2).`

Previously, NeoMutt used to warn with `Mailbox externally modified` and just work.

**Comment by mgedmin** [comment]
> Changes (moves / deletes) done via another client, then sync'd to local maildir by offlineimap/imapfilter, are not picked up by NeoMutt.

Also, new messages created locally with neomutt then synchronized to a remote maildir by offlineimap make NeoMutt show a duplicate entry of the same message.  This  can then lead to data loss it the user accidentally deletes the wrong "duplicate" (#2285).   This probably only applies to +record/+postponed.

I suspect that changing message flags via another instance of neomutt may cause the same "duplicate" message problem: when a filename in a maildir changes, neomutt notices and shows the message of the new filename, but it doesn't remove the message of the old filename.

**Comment by gahr** [comment]
This one puzzles me:

1. NeoMutt creates a new message file in the local maildir
2. offlineimap propagates the message to the IMAP server

What then?

3. offlineimap thinks there's a new message and fetches it as a new file
4. offlineimap removes the file that was generated by NeoMutt in step 1

or perhaps

3. offlineimap renames the file generated by NeoMutt in step 1

**Comment by mgedmin** [comment]
> offlineimap renames the file generated by NeoMutt in step 1

I think this is exactly what happens.  I tend to run offlineimap as `git add . && git commit -m mail && offlineimap && git add . && git commit -m mail`, and I see git commits that look like this: https://github.com/neomutt/neomutt/issues/2285#issuecomment-618249249

**Comment by gahr** [comment]
Right, so if offlineimap does his job before / after NeoMutt's sync cycle, at the next cycle NeoMutt will not find the old email, assume it's been deleted, and find a new one. All should be good.

However, if offlineimap goes into business while NeoMutt is sync'ing the mailbox, then I think we're out of luck. Could this be the case?

**Comment by lonerr** [comment]
The case is easy to reproduce without involving offlineimap at all.
Keeping mutt index with in one terminal window while removing message file with `rm` in another one is enough to trigger the issue.

**Comment by gahr** [comment]
Thanks for the simplified workflow - I could reproduce it :)

**Comment by gahr** [comment]
Good, I got `Mailbox was externally modified.  Flags may be wrong.`, but the message corresponding to the file I mv'd is still in the index.

**Comment by lonerr** [comment]
It certainly works better for me but still differently than previously.
In neomutt-20180716 externally removed message disappeared from index.
In 20200320 (with above patch) such a message still sits in index but marked with 'D' flag and disappears only after the next sync ($).
I'm not sure which behaviour is better though.

**Comment by gahr** [comment]
>  I'm not sure which behaviour is better though.

Let's pretend is the new one :)

**Comment by lonerr** [comment]
Honestly, I tend to agree :)
